### PR TITLE
Deprecate all interfaces

### LINKED ISSUE #2287: Support isinstance checks through ABCMeta for checkers and reporters
Currently the PyLinter categorizes checkers through the `__implements__` class attribute. This is not very standard, and is one more barrier for others to write plugins.

I propose

* Changing the interfaces to have a `ABCMeta` metaclass
* Change PyLinter to do isinstance checks to categorize checkers/reporters in addition to keeping the old `__implements__` checks

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Closes #2287.

### CODE REVIEW

File: pylint/lint/pylinter.py [hunk 1]
```diff
@@ -874,25 +874,45 @@ def _astroid_module_checker(self):
...
-                c
-                for c in _checkers
-                if interfaces.implements(c, interfaces.IRawChecker)
-                or isinstance(c, checkers.BaseRawFileChecker)
-            ]
+                    and c not in tokencheckers
+                    and c is not self
+                ):
+                    tokencheckers.append(c)  # pragma: no cover
+                    warnings.warn(
```
Comment (DanielNoord) [inline pylint/lint/pylinter.py:892]:
```suggestion
                    warnings.warn(. # pragma: no cover
```

File: pylint/lint/pylinter.py [hunk 2]
```diff
@@ -874,25 +874,45 @@ def _astroid_module_checker(self):
...
+        # TODO: 3.0: Remove deprecated if-statement # pylint: disable=fixme
+        for c in _checkers:
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", category=DeprecationWarning)
+                if (
+                    interfaces.implements(c, interfaces.IRawChecker)
+                    and c not in rawcheckers
+                ):
+                    rawcheckers.append(c)  # pragma: no cover
+                    warnings.warn(
```
Comment (DanielNoord) [inline pylint/lint/pylinter.py:910]:
```suggestion
                    warnings.warn(  # pragma: no cover
```

File: pylint/checkers/base_checker.py [hunk 3]
```diff
@@ -166,8 +173,14 @@ def create_message_definition_from_tuple(self, msgid, msg_tuple):
...
-            # TODO: Interfaces: Deprecate looking for implements here # pylint: disable=fixme
+            # TODO: 3.0: Remove deprecated if-statement # pylint: disable=fixme
             elif implements(self, (IRawChecker, ITokenChecker)):
+                warnings.warn(  # pragma: no cover
+                    "Checkers should subclass BaseTokenChecker or BaseRawFileChecker"
+                    "instead of using the __implements__ mechanism. Use of __implements__"
+                    "will no longer be supported in pylint 3.0",
+                    DeprecationWarning,
+                )
                 default_scope = WarningScope.LINE
```
Comment (DanielNoord) [inline pylint/checkers/base_checker.py:184]:
```suggestion
                default_scope = WarningScope.LINE  # pragma: no cover
```

File: pylint/lint/pylinter.py [hunk 4]
```diff
@@ -874,25 +874,45 @@ def _astroid_module_checker(self):
...
-                c
-                for c in _checkers
-                if interfaces.implements(c, interfaces.IRawChecker)
-                or isinstance(c, checkers.BaseRawFileChecker)
-            ]
+                    and c not in tokencheckers
+                    and c is not self
+                ):
+                    tokencheckers.append(c)  # pragma: no cover
+                    warnings.warn(. # pragma: no cover
```
Comment (DanielNoord) [inline pylint/lint/pylinter.py:892]:
```suggestion
                    warnings.warn(  # pragma: no cover
```

File: doc/whatsnew/2.14.rst [hunk 5]
```diff
@@ -169,6 +169,13 @@ Other Changes
 
   Ref #2287
 
+* All ``Interface`` classes in ``pylint.interfaces`` have been deprecated. You can subclass
+  the respective normal classes to get the same behaviour. The ``__implements__`` functionality
+  was based on a rejected PEP from 2001:
```
Comment (Pierre-Sassoulas) [inline doc/whatsnew/2.14.rst:174]: 👌

**Review by Pierre-Sassoulas — APPROVED** [review]
So much deprecation for 3.0 ! But removing pep that has been dead for 20 years is valuable, and pylint will benefit from being converted to modern python.

**Pull request merged**